### ISSUE TITLE
Compiler bug while transitioning to 0.19

### ISSUE THREAD

**Issue by silverweed** [crystal-lang/shards#120]
Hello, I was upgrading my code to work with 0.19 release (basically substituting all `as` with `.as`), and I stumbled upon a compiler bug:

```
Bug: trying to downcast SF::Texture+ <- SF::Texture
[8932062] *raise<Exception>:NoReturn +78
[8931863] ???
[17999049] ???
[17916263] ???
[18365280] *Crystal::CodeGenVisitor#visit<Crystal::Cast>:Bool +1296
[18030551] *Crystal::ASTNode+ +439
[18234360] *Crystal::CodeGenVisitor#prepare_call_args_non_external<Crystal::Call, Crystal::Def+, Crystal::Type+>:Tuple(Array(LLVM::Value), Bool) +856
[18218310] *Crystal::CodeGenVisitor#prepare_call_args<Crystal::Call, Crystal::Type+>:Tuple(Array(LLVM::Value), Bool) +118
[18197470] *Crystal::CodeGenVisitor#visit<Crystal::Call>:Bool +206
[18031171] *Crystal::ASTNode+ +1059
[18157946] *Crystal::CodeGenVisitor#accept_control_expression<Crystal::Return>:Crystal::Type+ +42
[18032713] *Crystal::ASTNode+ +2601
[18031356] *Crystal::ASTNode+ +1244
[18527785] *Crystal::CodeGenVisitor#visit<Crystal::ExceptionHandler>:Bool +553
[18033867] *Crystal::ASTNode+ +3755
[18031356] *Crystal::ASTNode+ +1244
[18118222] *Crystal::CodeGenVisitor#codegen_fun<String, Crystal::Def+, Crystal::Type+, Bool, LLVM::Module, Bool, Bool>:LLVM::Function +1678
[18138565] *Crystal::CodeGenVisitor#target_def_fun<Crystal::Def+, Crystal::Type+>:LLVM::Function +213
[18248533] *Crystal::CodeGenVisitor#codegen_call<Crystal::Call, Crystal::Def+, Crystal::Type+, Array(LLVM::Value)>:(Bool | LLVM::Value | Nil) +133
[18197692] *Crystal::CodeGenVisitor#visit<Crystal::Call>:Bool +428
[18031171] *Crystal::ASTNode+ +1059
[18161151] *Crystal::CodeGenVisitor#visit<Crystal::Assign>:(Bool | Nil) +239
[18031997] *Crystal::ASTNode+ +1885
[18527785] *Crystal::CodeGenVisitor#visit<Crystal::ExceptionHandler>:Bool +553
[18033867] *Crystal::ASTNode+ +3755
[18031356] *Crystal::ASTNode+ +1244
[18354718] *Crystal::CodeGenVisitor#codegen_if_branch<Crystal::CodeGenVisitor::Phi, Crystal::ASTNode+, LLVM::BasicBlock, Bool>:(LLVM::Value | Nil) +62
[18350102] *Crystal::CodeGenVisitor#visit<Crystal::If>:Bool +1078
[18031950] *Crystal::ASTNode+ +1838
[18031356] *Crystal::ASTNode+ +1244
[18118222] *Crystal::CodeGenVisitor#codegen_fun<String, Crystal::Def+, Crystal::Type+, Bool, LLVM::Module, Bool, Bool>:LLVM::Function +1678
[18138565] *Crystal::CodeGenVisitor#target_def_fun<Crystal::Def+, Crystal::Type+>:LLVM::Function +213
[18248533] *Crystal::CodeGenVisitor#codegen_call<Crystal::Call, Crystal::Def+, Crystal::Type+, Array(LLVM::Value)>:(Bool | LLVM::Value | Nil) +133
[18197692] *Crystal::CodeGenVisitor#visit<Crystal::Call>:Bool +428
[18031171] *Crystal::ASTNode+ +1059
[18031356] *Crystal::ASTNode+ +1244
[18118222] *Crystal::CodeGenVisitor#codegen_fun<String, Crystal::Def+, Crystal::Type+, Bool, LLVM::Module, Bool, Bool>:LLVM::Function +1678
[18138565] *Crystal::CodeGenVisitor#target_def_fun<Crystal::Def+, Crystal::Type+>:LLVM::Function +213
[18248533] *Crystal::CodeGenVisitor#codegen_call<Crystal::Call, Crystal::Def+, Crystal::Type+, Array(LLVM::Value)>:(Bool | LLVM::Value | Nil) +133
[18197692] *Crystal::CodeGenVisitor#visit<Crystal::Call>:Bool +428
[18031171] *Crystal::ASTNode+ +1059
[18234360] *Crystal::CodeGenVisitor#prepare_call_args_non_external<Crystal::Call, Crystal::Def+, Crystal::Type+>:Tuple(Array(LLVM::Value), Bool) +856
[18218310] *Crystal::CodeGenVisitor#prepare_call_args<Crystal::Call, Crystal::Type+>:Tuple(Array(LLVM::Value), Bool) +118
[18197470] *Crystal::CodeGenVisitor#visit<Crystal::Call>:Bool +206
[18031171] *Crystal::ASTNode+ +1059
[18354718] *Crystal::CodeGenVisitor#codegen_if_branch<Crystal::CodeGenVisitor::Phi, Crystal::ASTNode+, LLVM::BasicBlock, Bool>:(LLVM::Value | Nil) +62
[18350102] *Crystal::CodeGenVisitor#visit<Crystal::If>:Bool +1078
[18031950] *Crystal::ASTNode+ +1838
[18354718] *Crystal::CodeGenVisitor#codegen_if_branch<Crystal::CodeGenVisitor::Phi, Crystal::ASTNode+, LLVM::BasicBlock, Bool>:(LLVM::Value | Nil) +62
[18350102] *Crystal::CodeGenVisitor#visit<Crystal::If>:Bool +1078
[18031950] *Crystal::ASTNode+ +1838
[18031356] *Crystal::ASTNode+ +1244
[18389714] *Crystal::CodeGenVisitor#visit<Crystal::Yield>:Bool +2354
[18031249] *Crystal::ASTNode+ +1137
[18389714] *Crystal::CodeGenVisitor#visit<Crystal::Yield>:Bool +2354
[18031249] *Crystal::ASTNode+ +1137
[18389714] *Crystal::CodeGenVisitor#visit<Crystal::Yield>:Bool +2354
[18031249] *Crystal::ASTNode+ +1137
[18389714] *Crystal::CodeGenVisitor#visit<Crystal::Yield>:Bool +2354
[18031249] *Crystal::ASTNode+ +1137
[18031356] *Crystal::ASTNode+ +1244
[18355405] *Crystal::CodeGenVisitor#visit<Crystal::While>:Bool +605
[18032431] *Crystal::ASTNode+ +2319
[18031356] *Crystal::ASTNode+ +1244
[18241872] *Crystal::CodeGenVisitor#codegen_call_with_block<Crystal::Call, Crystal::Block, Crystal::Type+, Array(LLVM::Value)>:LLVM::Value +1072
[18197719] *Crystal::CodeGenVisitor#visit<Crystal::Call>:Bool +455
[18031171] *Crystal::ASTNode+ +1059
[18241872] *Crystal::CodeGenVisitor#codegen_call_with_block<Crystal::Call, Crystal::Block, Crystal::Type+, Array(LLVM::Value)>:LLVM::Value +1072
[18197719] *Crystal::CodeGenVisitor#visit<Crystal::Call>:Bool +455
[18031171] *Crystal::ASTNode+ +1059
[18031356] *Crystal::ASTNode+ +1244
[18241872] *Crystal::CodeGenVisitor#codegen_call_with_block<Crystal::Call, Crystal::Block, Crystal::Type+, Array(LLVM::Value)>:LLVM::Value +1072
[18197719] *Crystal::CodeGenVisitor#visit<Crystal::Call>:Bool +455
[18031171] *Crystal::ASTNode+ +1059
[18354718] *Crystal::CodeGenVisitor#codegen_if_branch<Crystal::CodeGenVisitor::Phi, Crystal::ASTNode+, LLVM::BasicBlock, Bool>:(LLVM::Value | Nil) +62
[18350072] *Crystal::CodeGenVisitor#visit<Crystal::If>:Bool +1048
[18031950] *Crystal::ASTNode+ +1838
[18031356] *Crystal::ASTNode+ +1244
[18241872] *Crystal::CodeGenVisitor#codegen_call_with_block<Crystal::Call, Crystal::Block, Crystal::Type+, Array(LLVM::Value)>:LLVM::Value +1072
[18197719] *Crystal::CodeGenVisitor#visit<Crystal::Call>:Bool +455
[18031171] *Crystal::ASTNode+ +1059
[18031356] *Crystal::ASTNode+ +1244
[18118222] *Crystal::CodeGenVisitor#codegen_fun<String, Crystal::Def+, Crystal::Type+, Bool, LLVM::Module, Bool, Bool>:LLVM::Function +1678
[18138565] *Crystal::CodeGenVisitor#target_def_fun<Crystal::Def+, Crystal::Type+>:LLVM::Function +213
[18248533] *Crystal::CodeGenVisitor#codegen_call<Crystal::Call, Crystal::Def+, Crystal::Type+, Array(LLVM::Value)>:(Bool | LLVM::Value | Nil) +133
[18197692] *Crystal::CodeGenVisitor#visit<Crystal::Call>:Bool +428
[18031171] *Crystal::ASTNode+ +1059
[18031356] *Crystal::ASTNode+ +1244
[18118222] *Crystal::CodeGenVisitor#codegen_fun<String, Crystal::Def+, Crystal::Type+, Bool, LLVM::Module, Bool, Bool>:LLVM::Function +1678
[18138565] *Crystal::CodeGenVisitor#target_def_fun<Crystal::Def+, Crystal::Type+>:LLVM::Function +213
[18248533] *Crystal::CodeGenVisitor#codegen_call<Crystal::Call, Crystal::Def+, Crystal::Type+, Array(LLVM::Value)>:(Bool | LLVM::Value | Nil) +133
[18197692] *Crystal::CodeGenVisitor#visit<Crystal::Call>:Bool +428
[18031171] *Crystal::ASTNode+ +1059
[18031356] *Crystal::ASTNode+ +1244
[18354718] *Crystal::CodeGenVisitor#codegen_if_branch<Crystal::CodeGenVisitor::Phi, Crystal::ASTNode+, LLVM::BasicBlock, Bool>:(LLVM::Value | Nil) +62
[18350102] *Crystal::CodeGenVisitor#visit<Crystal::If>:Bool +1078
[18031950] *Crystal::ASTNode+ +1838
[18354718] *Crystal::CodeGenVisitor#codegen_if_branch<Crystal::CodeGenVisitor::Phi, Crystal::ASTNode+, LLVM::BasicBlock, Bool>:(LLVM::Value | Nil) +62
[18350102] *Crystal::CodeGenVisitor#visit<Crystal::If>:Bool +1078
[18031950] *Crystal::ASTNode+ +1838
[18031356] *Crystal::ASTNode+ +1244
[18118222] *Crystal::CodeGenVisitor#codegen_fun<String, Crystal::Def+, Crystal::Type+, Bool, LLVM::Module, Bool, Bool>:LLVM::Function +1678
[18109686] *Crystal::CodeGenVisitor#visit<Crystal::ProcLiteral>:Bool +150
[18032619] *Crystal::ASTNode+ +2507
[18354718] *Crystal::CodeGenVisitor#codegen_if_branch<Crystal::CodeGenVisitor::Phi, Crystal::ASTNode+, LLVM::BasicBlock, Bool>:(LLVM::Value | Nil) +62
[18350072] *Crystal::CodeGenVisitor#visit<Crystal::If>:Bool +1048
[18031950] *Crystal::ASTNode+ +1838
[18354718] *Crystal::CodeGenVisitor#codegen_if_branch<Crystal::CodeGenVisitor::Phi, Crystal::ASTNode+, LLVM::BasicBlock, Bool>:(LLVM::Value | Nil) +62
[18350102] *Crystal::CodeGenVisitor#visit<Crystal::If>:Bool +1078
[18031950] *Crystal::ASTNode+ +1838
[18354718] *Crystal::CodeGenVisitor#codegen_if_branch<Crystal::CodeGenVisitor::Phi, Crystal::ASTNode+, LLVM::BasicBlock, Bool>:(LLVM::Value | Nil) +62
[18350102] *Crystal::CodeGenVisitor#visit<Crystal::If>:Bool +1078
[18031950] *Crystal::ASTNode+ +1838
[18118222] *Crystal::CodeGenVisitor#codegen_fun<String, Crystal::Def+, Crystal::Type+, Bool, LLVM::Module, Bool, Bool>:LLVM::Function +1678
[18138565] *Crystal::CodeGenVisitor#target_def_fun<Crystal::Def+, Crystal::Type+>:LLVM::Function +213
[18248533] *Crystal::CodeGenVisitor#codegen_call<Crystal::Call, Crystal::Def+, Crystal::Type+, Array(LLVM::Value)>:(Bool | LLVM::Value | Nil) +133
[18197692] *Crystal::CodeGenVisitor#visit<Crystal::Call>:Bool +428
[18031171] *Crystal::ASTNode+ +1059
[18044791] *Crystal::CodeGenVisitor#visit<Crystal::TupleLiteral>:Bool +311
[18031903] *Crystal::ASTNode+ +1791
[18234360] *Crystal::CodeGenVisitor#prepare_call_args_non_external<Crystal::Call, Crystal::Def+, Crystal::Type+>:Tuple(Array(LLVM::Value), Bool) +856
[18218310] *Crystal::CodeGenVisitor#prepare_call_args<Crystal::Call, Crystal::Type+>:Tuple(Array(LLVM::Value), Bool) +118
[18197470] *Crystal::CodeGenVisitor#visit<Crystal::Call>:Bool +206
[18031171] *Crystal::ASTNode+ +1059
[18031356] *Crystal::ASTNode+ +1244
[18389714] *Crystal::CodeGenVisitor#visit<Crystal::Yield>:Bool +2354
[18031249] *Crystal::ASTNode+ +1137
[18031356] *Crystal::ASTNode+ +1244
[18241872] *Crystal::CodeGenVisitor#codegen_call_with_block<Crystal::Call, Crystal::Block, Crystal::Type+, Array(LLVM::Value)>:LLVM::Value +1072
[18197719] *Crystal::CodeGenVisitor#visit<Crystal::Call>:Bool +455
[18031171] *Crystal::ASTNode+ +1059
[18031356] *Crystal::ASTNode+ +1244
[18118222] *Crystal::CodeGenVisitor#codegen_fun<String, Crystal::Def+, Crystal::Type+, Bool, LLVM::Module, Bool, Bool>:LLVM::Function +1678
[18138565] *Crystal::CodeGenVisitor#target_def_fun<Crystal::Def+, Crystal::Type+>:LLVM::Function +213
[18248533] *Crystal::CodeGenVisitor#codegen_call<Crystal::Call, Crystal::Def+, Crystal::Type+, Array(LLVM::Value)>:(Bool | LLVM::Value | Nil) +133
[18197692] *Crystal::CodeGenVisitor#visit<Crystal::Call>:Bool +428
[18031171] *Crystal::ASTNode+ +1059
[18364031] *Crystal::CodeGenVisitor#visit<Crystal::Cast>:Bool +47
[18030551] *Crystal::ASTNode+ +439
[18161151] *Crystal::CodeGenVisitor#visit<Crystal::Assign>:(Bool | Nil) +239
[18031997] *Crystal::ASTNode+ +1885
[18031356] *Crystal::ASTNode+ +1244
[18118222] *Crystal::CodeGenVisitor#codegen_fun<String, Crystal::Def+, Crystal::Type+, Bool, LLVM::Module, Bool, Bool>:LLVM::Function +1678
[18138565] *Crystal::CodeGenVisitor#target_def_fun<Crystal::Def+, Crystal::Type+>:LLVM::Function +213
[18248533] *Crystal::CodeGenVisitor#codegen_call<Crystal::Call, Crystal::Def+, Crystal::Type+, Array(LLVM::Value)>:(Bool | LLVM::Value | Nil) +133
[18197692] *Crystal::CodeGenVisitor#visit<Crystal::Call>:Bool +428
[18031171] *Crystal::ASTNode+ +1059
[18031356] *Crystal::ASTNode+ +1244
[18118222] *Crystal::CodeGenVisitor#codegen_fun<String, Crystal::Def+, Crystal::Type+, Bool, LLVM::Module, Bool, Bool>:LLVM::Function +1678
[18138565] *Crystal::CodeGenVisitor#target_def_fun<Crystal::Def+, Crystal::Type+>:LLVM::Function +213
[18248533] *Crystal::CodeGenVisitor#codegen_call<Crystal::Call, Crystal::Def+, Crystal::Type+, Array(LLVM::Value)>:(Bool | LLVM::Value | Nil) +133
[18197692] *Crystal::CodeGenVisitor#visit<Crystal::Call>:Bool +428
[18031171] *Crystal::ASTNode+ +1059
[18161151] *Crystal::CodeGenVisitor#visit<Crystal::Assign>:(Bool | Nil) +239
[18031997] *Crystal::ASTNode+ +1885
[18031356] *Crystal::ASTNode+ +1244
[18118222] *Crystal::CodeGenVisitor#codegen_fun<String, Crystal::Def+, Crystal::Type+, Bool, LLVM::Module, Bool, Bool>:LLVM::Function +1678
[18138565] *Crystal::CodeGenVisitor#target_def_fun<Crystal::Def+, Crystal::Type+>:LLVM::Function +213
[18248533] *Crystal::CodeGenVisitor#codegen_call<Crystal::Call, Crystal::Def+, Crystal::Type+, Array(LLVM::Value)>:(Bool | LLVM::Value | Nil) +133
[18197692] *Crystal::CodeGenVisitor#visit<Crystal::Call>:Bool +428
[18031171] *Crystal::ASTNode+ +1059
[18031356] *Crystal::ASTNode+ +1244
[18118222] *Crystal::CodeGenVisitor#codegen_fun<String, Crystal::Def+, Crystal::Type+, Bool, LLVM::Module, Bool, Bool>:LLVM::Function +1678
[18138565] *Crystal::CodeGenVisitor#target_def_fun<Crystal::Def+, Crystal::Type+>:LLVM::Function +213
[18248533] *Crystal::CodeGenVisitor#codegen_call<Crystal::Call, Crystal::Def+, Crystal::Type+, Array(LLVM::Value)>:(Bool | LLVM::Value | Nil) +133
[18197692] *Crystal::CodeGenVisitor#visit<Crystal::Call>:Bool +428
[18031171] *Crystal::ASTNode+ +1059
[18161151] *Crystal::CodeGenVisitor#visit<Crystal::Assign>:(Bool | Nil) +239
[18031997] *Crystal::ASTNode+ +1885
[18031356] *Crystal::ASTNode+ +1244
[17717055] *Crystal::Compiler#codegen<Crystal::Program, Crystal::ASTNode+, Array(Crystal::Compiler::Source), String>:(Array(String) | Nil) +1327
[11986866] *Crystal::Command#build:Crystal::Compiler::Result +162
[9497719] *Crystal::Command#run:(Array(Crystal::ImplementationTrace) | Array(Crystal::Init::View+:Class) | Array(String) | Bool | Crystal::Compiler::Result | Hash(String, String) | IO::FileDescriptor | Nil) +1591
[9058756] main +26212
[140595562750609] __libc_start_main +241
[8902474] _start +42
[0] ???

Error: you've found a bug in the Crystal compiler. Please open an issue, including source code that will allow us to reproduce the bug: https://github.com/crystal-lang/crystal/issues
```

The source code can be found [here](https://github.com/silverweed/lifish-edit).
My project has [crsfml](https://github.com/BlaXpirit/crsfml) as dependency, plus a small C library I link externally: [nfd](https://github.com/mlabbe/nativefiledialog).

All the object files except the main (`src/LifishEdit.cr`) compile without errors.

**Comment by asterite** [comment]
@silverweed Thanks!

I moved this issue to the Crystal repository, and reduced it: https://github.com/crystal-lang/crystal/issues/3304

**Comment by asterite** [comment]
@silverweed By the way, you are using too many casts (`as`) in your code.

For example in this line:

```cr
return @textures[key] = SF::Texture.from_file(LE::Utils.get_graphic(key)).as SF::Texture
```

remove the last cast `.as SF::Texture` just works. I don't know why you put that.

Other cases: in `class App` you have:

```cr
	def font
		@font.as SF::Font
	end

	def menu
		@menu.as LE::Menu
	end

	def ls
		@ls.as LE::LevelSet
	end
```

But what you really want is just asserting that these are not `nil`, not writing the type name again. This works:

```cr
	def font
		@font.not_nil!
	end

	def menu
		@menu.not_nil!
	end

	def ls
		@ls.not_nil!
	end
```

In fact, this case where you know an instance variable must be initialized at that point is so common that there's a macro to help: [getter!](https://crystal-lang.org/api/0.19.1/Object.html#getter%21%28%2Anames%29-macro). So the above can be written as:

```cr
    getter! font
    getter! menu
    getter! ls
```

In short: use `not_nil!` instead of an explicit cast, and use `getter!` and `property!` macros. This should make your coding experience a bit more pleasant, I think.

**Comment by asterite** [comment]
@silverweed Finally, if you remove those extra casts, code should compile just fine because the bug involved casting a type to itself, with that type having subtypes. Removing [this cast](https://github.com/silverweed/lifish-edit/blob/master/src/LifishEdit/cache.cr#L15) makes your app compile again.

**Comment by silverweed** [comment]
Thank you very much!
All those casts derive either from the time when I used an "ancient" version of the compiler (when I started the project Crystal was like at 0.5) or (more likely) from my inexpertise with the language :D

I have yet to delve deep into the API docs to discover all those useful macros & methods; I guess I should pass more time on IRC to ask suggestions :)
